feature: lock_timeout can be provided when making a payment

offerm opened the pull request:
Example:
```
curl -X POST http://localhost:5001/api/v1/payments/0xF6f11bb624aabb2782FD53CcbF10db73f303F367/0x22b001098c212eA447EF291dE687Bf5C2Ef40914 -H 'Content-Type: application/json' --data-raw '{"amount": 500, "identifier": 987654326, "secret": "0x78c8d676e2f2399aa2a015f3433a2083c55003591a0f3f3349b6e50fc9ca44f1", "lock_timeout": 50}'
```
for a channel to be used, the lock_timeout should be bigger than reveal_timeout and not bigger than settle_timeout.


Fixes #4669

offerm commented:
I'm working to provide some tests for this feature.  Looking forward to get initial feedback.

hackaugusto commented on raiden/constants.py:
```diff
@@ -57,6 +58,7 @@
 ZERO_TOKENS = TokenAmount(0)
 
 ABSENT_SECRET = Secret(b"")
+ABSENT_BLOCKTIMEOUT = BlockTimeout(0)
```
Could you please revert this? A block timeout of `0` is most likely invalid. For the case of a lock timeout it will always create an invalid lock expiration, if we do instantiate this value the type checker cannot help us find bugs.

hackaugusto commented on raiden/raiden_service.py:
```diff
@@ -1117,6 +1127,7 @@ def start_mediated_transfer_with_secret(
         identifier: PaymentID,
         secret: Secret,
         secrethash: SecretHash = None,
+        locktimeout: BlockTimeout = ABSENT_BLOCKTIMEOUT,
```
IMO we should maintain the type `Optional[BlockTimeout]` here. And if the value `is None` then `get_safe_initial_expiration` is called, otherwise we use the user's provided value.

offerm replied:
@hackaugusto 

Initially I just used `None` for missing `locktimeout`. This works across the code but fail with restoring `TransferDescriptionWithSecretState` from the database. Looks like the `dataclass` is not ready to decode None value.

For that reason I had to use a value when locktimeout is not given. 

I will create a new type for locktimeout (not to be confused with blocktimeout) and create an ABSENT_LOCKTIMEOUT with a value of 0.

Ok?

rakanalh replied:
> Initially I just used None for missing locktimeout. This works across the code but fail with restoring TransferDescriptionWithSecretState from the database. Looks like the dataclass is not ready to decode None value.

I can help you with the restoration if i know exactly what the problem is.

hackaugusto commented on raiden/transfer/channel.py:
```diff
@@ -190,18 +200,18 @@ def is_channel_usable_for_mediation(
-    lock_timeout_valid = (
-        lock_timeout > 0
-        and channel_state.settle_timeout >= lock_timeout
-        and channel_state.reveal_timeout < lock_timeout
```
These checks are important. A lock that does not respect these constraints is not safe to be used. For more details check the function docstring.

offerm replied:
This is checked as part of `is_channel_usable_for_new_transfer` which is being called by `is_channel_usable_for_mediation` and also in the case of initiator.

In the initiator case the value can be None but in case of mediation it must be a positive number within the allowed range.

hackaugusto commented on raiden/transfer/channel.py:
```diff
@@ -217,12 +227,17 @@ def is_channel_usable_for_new_transfer(
     """
     pending_transfers = get_number_of_pending_transfers(channel_state.our_state)
     distributable = get_distributable(channel_state.our_state, channel_state.partner_state)
+    lock_timeout_valid = lock_timeout == ABSENT_BLOCKTIMEOUT or (
```
this check is good, I think it should only change to `lock_timeout is None or`

offerm commented on raiden/api/python.py:
```diff
@@ -868,6 +868,7 @@ def transfer_and_wait(
         transfer_timeout: int = None,
         secret: Secret = None,
         secrethash: SecretHash = None,
+        locktimeout: BlockTimeout = None,
```
@rakanalh I did the same thing as done for secret_hash.... 
When I added secret_hash you asked me to have secret_hash on the json level and secrethash internally. So I did the exact same thing here.

Anyway, I will change to lock_timeout across the board.

rakanalh replied:
Sure, `secrethash` is the only instance where it's used a single word. My point is to have consistency across the codebase.

offerm commented on raiden/api/rest.py:
```diff
@@ -1022,6 +1024,7 @@ def initiate_payment(
                 identifier=identifier,
                 secret=secret,
                 secrethash=secret_hash,
+                locktimeout=lock_timeout,
```
see `locktimeout` one line above....

offerm commented:
@rakanalh @hackaugusto I renamed locktimeout to lock_timeout across the code and added tests. 
Also remove the ABSENT_BLOCKTIMEOUT.
Everything works fine but lint is complaining about using None with `BlockTimeout` type.

Let me know what you remand to do.

Also, the added tests has some ugly constants. Let me know the base way to find the channel between the peers so I can fetch and use the `settle_timeout` and `reveal_timeout`.

rakanalh commented:
@offerm In this case, we declare lock_timeout as `Optional[BlockTimeout]` in the transfer description. I did that in your PR and pushed the change.

Now during linting, there remains:

```
raiden/transfer/mediated_transfer/initiator.py:242: error: Argument "lock_timeout" to "is_channel_usable_for_new_transfer" has incompatible type "Optional[BlockTimeout]"; expected "BlockTimeout"
```
which means you have to decide how you want to handle None lock timeout in that function, i'll leave that up to you.

offerm commented:
@rakanalh can you please check why the test fails? It is working for me locally.

The test is using some constants - when is the best way to replace these with the channel values?

rakanalh commented on raiden/api/v1/encoding.py:
```diff
@@ -338,6 +338,7 @@ class PaymentSchema(BaseSchema):
     identifier = fields.Integer(missing=None)
     secret = SecretField(missing=None)
     secret_hash = SecretHashField(missing=None)
+    lock_timeout = fields.Integer(missing=None)
```
Please add a documentation for this in the [`Payment` section](https://github.com/raiden-network/raiden/blob/develop/docs/rest_api.rst#payments) of the docs:

rakanalh commented:
@offerm I checked the failing tests and here's the summary:

In the logs of the failing tests i was able to spot:
```
errors = Payment couldn't be completed because: lock expired status_code = 409
```

Although the target received `ActionInitTarget` 1 second after the transfer was initiated, the target checks if the remaining blocks are enough for reveal:

https://github.com/raiden-network/raiden/blob/a35698a550f738336bebcefa02f1a2a63504e5e2/raiden/transfer/mediated_transfer/target.py#L122

https://github.com/raiden-network/raiden/blob/a35698a550f738336bebcefa02f1a2a63504e5e2/raiden/transfer/mediated_transfer/mediator.py#L104-L109

And since the remaining blocks are less than the `reveal_timeout`, the target will reject the transfer and decides not to ask for a secret as it sees that the transfer is about to expire.

The reason this fails on CI is because CI is generally slower than regular machines. So it's best to choose a lock timeout value that is sufficient.

In the tests, we have 2 nodes and this number affects the default reveal timeout we use in tests:

```
return number_of_nodes * 4 + DEFAULT_NUMBER_OF_BLOCK_CONFIRMATIONS
```
which means that the test channels reveal timeout value would be 13.

I would recommend using lock timeout values which are at least twice the reveal timeout. We have that documented here:

https://github.com/raiden-network/raiden/blob/a35698a550f738336bebcefa02f1a2a63504e5e2/raiden/transfer/channel.py#L133-L148

offerm commented:
@rakanalh ready?

rakanalh commented on raiden/transfer/channel.py:
```diff
@@ -219,19 +222,26 @@ def is_channel_usable_for_new_transfer(
+    if lock_timeout and (
+        lock_timeout > channel_state.settle_timeout or lock_timeout <= channel_state.reveal_timeout
+    ):
+        lock_timeout_valid = False
```
These 4 lines can be replaced with
```
lock_timeout_valid = (
    lock_timeout <= channel_state.settle_timeout and
    lock_timeout > channel_state.reveal_timeout
)
```

Also, my suggestion changes the comparison operators to match the conditions you've removed from the caller function

offerm replied:
You version gives this:
```
(raiden) admins-MacBook-Pro-2:raiden admin$ make lint
mypy raiden
raiden/transfer/channel.py:239: error: Unsupported operand types for >= ("BlockTimeout" and "None")
raiden/transfer/channel.py:239: note: Left operand is of type "Optional[BlockTimeout]"
raiden/transfer/channel.py:240: error: Unsupported operand types for < ("BlockTimeout" and "None")
raiden/transfer/channel.py:240: note: Left operand is of type "Optional[BlockTimeout]"
make: *** [mypy] Error 1
```

changing it to:
```
    lock_timeout_valid = lock_timeout is None or (
        lock_timeout <= channel_state.settle_timeout and
        lock_timeout > channel_state.reveal_timeout
    )
```
solved the lint problem but my IDE (PyCharm) comlains "Simplifu chained comparison"

I changed to the second version but it looks buggy.

rakanalh commented on raiden/api/python.py:
```diff
@@ -973,6 +976,7 @@ def transfer_async(
             identifier=identifier,
             secret=secret,
             secrethash=secrethash,
+            lock_timeout=lock_timeout,
```
This should be validated prior sending it through the state machine.
```
lock_timeout_valid = (
    lock_timeout <= channel_state.settle_timeout and
    lock_timeout > channel_state.reveal_timeout
)
```
Copied from `is_channel_usable`.

We should make sure the value is valid first.

Also, please add a regression test for this to make sure invalid values are rejected.

offerm replied:
How can this be validated here without a channel in hand?
also, how do you define an invalid value?  

Note that there are test cases that covers invalid values (values for which there are no route to target).

rakanalh replied:
> How can this be validated here without a channel in hand?

You could use 
```
    channel_state = views.get_channelstate_for(
            chain_state=chain_state,
            token_network_registry_address=registry_address,
            token_address=token_address,
            partner_address=partner_address,
        )
```

> also, how do you define an invalid value?

You can see that `transfer_async` function has a set of validations raising `InvalidXError`. So you could introduce this new exception `InvalidLockTimeoutError` and raise it if the value is invalid.

> Note that there are test cases that covers invalid values (values for which there are no route to target).

Right, so we can use those existing tests to add one more check for the invalid lock timeout.

rakanalh commented on docs/rest_api.rst:
```diff
@@ -683,6 +683,8 @@ Payments
 
    :reqjson int amount: Amount to be sent to the target
    :reqjson int identifier: Identifier of the payment (optional)
+   :reqjson int lock_timeout: lock timeout, in blocks, to be used with the payment. Default is 2 * channel's reveal_timeout (optional)
```
Maybe add a note below talking about the limitations of the value provided. You do mention what the default is, but it's not clear why the value has to be larger than the reveal timeout. Please clarify it in a note below this line.

offerm replied:
Added that to the lock_timeout description. I prefer not to make changes to the .rst as PyCharm fails to display the result properly. If important, please help with this note.

Thanks @rakanalh

offerm commented:
@rakanalh please review.

offerm commented:
> The PR looks good. However, i am going to approve it after adding the validation in the `python.py` module and adding the regression test for this validation.

@rakanalh this does not make sense to me. You assume that you have a direct channel to the `target`. This does not have to be the case. There can also be multiple routes between the `initiator` and the `target`.

Even if there is a direct channel, this check may fail while there is a valid route to the tager via some other nodes.

Let me know what you think.

rakanalh commented:
You're actually right... i did not think this thoroughly.

rakanalh commented:
Alright so i rebased the PR and will merge once the tests are green.
